### ISSUE TITLE
grammar not loading

### ISSUE THREAD

**Issue by balmas** [alpheios-project/safari-app-front#3]
I didn't have a chance to debug this through to find the cause, but the grammar resource isn't loading in the grammar frame. Maybe a Safari restriction in iFrames?

**Comment by irina060981** [comment]
After some investigating - I have found - that I have commented it upon development process and forgot about it.
I have commented it - because there was a problem with iframe - that I didn't solved - it creates an infinitive cycle for the event `DOMContentLoaded` - and everytime iframe is loaded - it restarts all content process initialization.
I will try to solve it again!

**Comment by balmas** [comment]
Ah, that's not too surprising. I ran into some problems with iFrames and loading events on Firefox too. Maybe you can check whether the event is triggered from an iframe in the DomConentLoaded handler and abort early?

**Comment by irina060981** [comment]
(simply coping here from slack)
I am trying to solve the issue with iframe for several hours - and I am really far from success 

1) I have tried to set `v-on:load` for iframe and define a method with `preventDefault`, `stopPropagation`, `stopImmediatePropagation`
2) I have tried to remove v-if from iframe and add `return this.res.url ? this.res.url : 'about:blank'`
3) I have tried to create global variable for `window`, `safari` to check if contentProcess was already initialized
4) I have tried to check document.body class, searching for `alpheios-panel`

as I could see from my experiments
DOMContentLoaded fires when iframe src property is changing / when it is appearing in HTML Layout
and document layout (to use body for checking) is loaded to document variable much later

**Comment by balmas** [comment]
@kirlat is going to take a look at this.

**Comment by kirlat** [comment]
I did multiple tests, combined that with whatever scarce information was available on the net, and here are my findings.

It seems to me that the problem arises from a difference between Safari App Extension and webextension architectures.

In webextension, a background script is loaded once extension is activated. Background script is running all the time in, well, background. It can keep track of webextension button and menu interactions on individual pages, or tabs, within a browser. When user selects an "activate" action, background scripts gets a notification about that from a webextension API along with a tab ID of a page where it happened. With this information in hand, background script can load (inject) a content script into a corresponding tab and activate that.

In Safari App Extension, an architecture is different. There is a native macOS containing app, that has a native Safari app extension code inside of it, and there are JS and CSS content script and style files (still not sure where are they stored physically - are they packed inside a containing app?). However, there is no code in an app extension that explicitly injects those script and style files into a containing page. Those CSS and JS files are listed in a `plist` config file only. I was not able to find any clear information on an injection process (info on Safary App Extension is really hard to get), but from my test it seems that JS and CSS files are injected automatically into every page that come into existence within Safari, including iframes and some other auxillary documents. Every nested document seems to be like a new webpage to the injection code. 

Here lies the problem. Our content script was designed to be injected only into pages where user decidedly activate Alpheios, and it was designed to be loaded only once per tab. In Safari, however, it is injected into every page, even if Alpheios is not activated there, and it is injected into any document that is nested within a page. Since during initialization our script create sub pages (grammar in an iframe), this triggers an infinite initialization loop.

As a solution to this, we can filter out pages that are not "root" (i.e. are nested) and skip content process initialization there. In the current build I used `event.currentTarget.URL` to filter URIs that contain `grammar.alpheios.net` (grammar in an iframe) and `about:blank` (this seems to be quirks of how Vue.js interacts with Safari for component rendering). That was working fine with me. As an alternative filtering mechanism, we can compare a current page with the `window.top`. I'm not sure which method be more reliable. Please let me know what do you think.

The other implication (but that's probably best addressed during refactoring) is that we should avoid doing any heavy lifting during content process initialization (as this will happen in Safari on every "root" page) and move that into an activation phase, when users indicated his/her intent to activate an extension. This will eliminate initialization overhead on pages where Alpheios never be activated.

@irina060981, what does your experience working with Safari App Extension tells you? Does the description above depicts correctly how a Safari App Extension works? What's your opinion on this? Thanks!

**Comment by irina060981** [comment]
Hello, @kirlat !

Some comments:

1) *still not sure where are they stored physically - are they packed inside a containing app?*
All injected files (js, css, icons and may be some other types of files) are registered in a specific configuration file - [Info.plist](https://github.com/alpheios-project/safari-app-back/blob/master/AlpheiosSafariV2/Info.plist)

It has various settings - and each target has its own info.plist file.
For example our project has 2 info.plist files - for application and for safari app extension

All injected files are laying in the folder of its target (in our case for Safari App Extension js, css, icons are laying inside AlpheiosSafari folder) and you could update it manually from XCode or replacing with a new version.

When you create a build all files are packed together and could be unpacked.
( I am not sure - may be I haven't understand your question clearly - so I decided to explain it all over)

2) *About activating*
    It seems to me that activating of safari app extension is working this way:
   
   1) you should activate application
   2) it loads some compiled files to the specific safari folder (I had to clear unsuccessfull copies and versions of it after some crushes)
   3) you could activate it from Extensions tab (only if it has compiled files in safari's folder and if a parent application is active)
   4) there are 3 objects in the safari application - Window (SFSafariWindow)  -> active Tab -> active Page (SFSafariPage)
       It has links only from window to page and not vice versa
       Messages to Background application could be sent only from one active page and there are no any identifications for this tree besides some hashes (that I used as tabs).
      So when the background gets the message - the only way to communicate with the front part - we know only page hash and nothing more - so there is no even way to identify the window.
       
        So if your hypothesis is right - (and it is really an interesting one!!! I haven't guess :) ), then the iframe has its own window  and I am not sure how could it be tracked in the background - I didn't made such tests. 

       But if to be honest it is really a strange and unsecure behaviour as iframes should be secured from external scripts and vice versa (at least I thought this way).
        
3) *About filtering*
   It seems to me that any iframe could create this circular reactivation (if your hypothis is right) so Are you sure that filtering by domain is enough?
  And filtering by position could has many variant as not every page should be sticked to the top of the window - on my mac it is really difficult to use browser streched over the window :)

**Comment by irina060981** [comment]
I will try to think and to do some tests tomorrow because it is really difficult to jump between safari app extension and inflection games - so I haven't ideas right now.

**Comment by kirlat** [comment]
Thanks for detailed explanation of Safari behavior, Irina! I think we should save it somewhere to the docs.

I was putting a debug statement into `safari-app-front\index.js` before the `DOMContentLoaded` callback (i.e. into the global context), and this statement was printed several times for each page. It looks like the content script is loaded several times, once for each document within a tab: once for the "root" document, once for each iframe within that root document, and if Vue.js creates some temporary inner documents during rendering, then for those ones too. So it's not that content script is loaded once per tab and then `DOMContentLoaded` callback fired several times, but it's rather content script loaded several times and `DOMContentLoaded` called once per each script load. That's why any global variable to store content process initialization state (i.e. if it was initialized already) does not work: each callback has its own isolated global environment (or it seems so).

So the only recourse, I thought, is to identify if the script is running within a "root" page, or within one of the nested pages. I used `event.currentTarget.URL` to identify nested pages (i.e. `grammar.alpheios.net` or `about: blank`) and skipped initialization there. Worked well for me, but needs testing in more different environment.

The other solution might be to compare `event.currentTarget` (the document script is currently running in) with `window.top` (the topmost document in tab hierarchy, i.e. the "root" document) and initialize content process only if we're at the highest hierarchy level. This latter solution seems to be more universal, but I'm afraid Safari might mess up page hierarchy information in some edge cases (maybe not?) and things will crumble.

**Comment by balmas** [comment]
This is very interesting.

If @kirlat is right that Alpheios gets injected into every page, then I think we will have to look at slimming down what happens upon content process activation. For example, I don't think we want to load the inflection tables into every page.  

I don't have any other ideas to offer immediately about how to prevent the injection or activation of the script for specific windows/frames. Will keep thinking about this too.

**Comment by balmas** [comment]
closing this. Will enter a new issue on the webextension for remaining changes for this.
